src/Microsoft.OData.Core/SelectedPropertiesNode.cs

```diff
@@ -441,11 +461,131 @@ internal IEnumerable<IEdmNavigationProperty> GetSelectedNavigationProperties(IEd
             return selectedNavigationProperties.Distinct();
         }
 
+        /// <summary>
+        /// Gets all selected structured properties at the current node.
+        /// </summary>
+        /// <param name="structuredType">The current structured type.</param>
+        /// <returns>The selected structured properties at this node level.</returns>
+        /// <remarks>Result does not include the dynamic properties, which can be retrieved by <code>GetSelectedDynamicProperties</code></remarks>
+        internal IEnumerable<IEdmProperty> GetSelectedProperties(IEdmStructuredType structuredType)
+        {
+            if (this.selectionType == SelectionType.Empty)
+            {
+                return EmptyEdmProperties;
+            }
+
+            // We cannot determine the selected properties without the user model. This means we won't be computing the missing properties.
+            if (structuredType == null)
+            {
+                return EmptyEdmProperties;
+            }
+
+            if (this.selectionType == SelectionType.EntireSubtree || this.hasWildcard)
+            {
+                return structuredType.Properties();
+            }
+
+            Debug.Assert(this.selectedProperties != null, "selectedProperties != null");
+
+            // Get properties selected, and filter out unrecognized properties.
+            return this.selectedProperties.Select(structuredType.FindProperty).OfType<IEdmProperty>();
+        }
+
+        /// <summary>
+        /// Return names of dynamic properties.
+        /// </summary>
+        /// <param name="structuredType">The current structured type.</param>
+        /// <returns>The names of dynamic properties.</returns>
+        internal IEnumerable<string> GetSelectedDynamicProperties(IEdmStructuredType structuredType)
+        {
+            IEnumerable<string> dynamicProperties = Enumerable.Empty<string>();
+
+            if (this.selectedProperties != null && this.selectedProperties.Count > 0)
+            {
+                dynamicProperties = this.selectedProperties.Select(
+                    prop => structuredType.FindProperty(prop) == null
+                    ? prop
+                    : null)
+                    .OfType<string>();
+            }
+
+            return dynamicProperties;
+        }
+
+        /// <summary>
+        /// Gets the list of names for selected declared or dynamic properties having values of null.
+        /// </summary>
+        /// <param name="structuredType">The current structured type.</param>
+        /// <param name="resourceState">The current resource state.</param>
+        /// <returns>The list of property names with null values.</returns>
+        internal IList<string> GetSelectedNullValueProperties(IEdmStructuredType structuredType,

```

Comment:
>GetSelectedNullValueProperties [](start = 31, length = 30)

We are doing a lot of the same work (i.e., determining the set of nullable properties) for each resource in the result.  I wonder if it wouldn't be more efficient to determine the set of nullable property names up front, and then, at the end of writing each resource, restore nulls for the properties from that static list that have not been read.

Reply:
The current logic follows the overall design of ODataJsonLightReader's EndEntry() method and has been the design from the beginning. It restores all missing nulls for the current entry all at once when the entry is deserialized. If we restored at the end of writing each resource, it would still need to perform the same restoration of null values, and with additional cost of caching information for null properties. Therefore doing it in flight is more efficient.

---
In reply to: [233190829](https://github.com/OData/odata.net/pull/1106#discussion_r233190829) [](ancestors = 233190829)

Reply:
I understand that we are restoring null values at the end -- that's fine.  What I am questioning is whether we have to identify the set of nullable properties each time, or if we can do that once for the result.

---
In reply to: [233299115](https://github.com/OData/odata.net/pull/1106#discussion_r233299115) [](ancestors = 233299115,233190829)

Reply:
Since SelectedPropertiesNode is designed as part of context url parsing and is used to represent relatively static context url information, I am not sure it is proper for SelectedPropertiesNode to cache null properties, which is related to runtime payload and is not intrinsic part of SelectedPropertiesNode.
Additionally, there are no 'recalculating' per se; the calculations of set of omitted null properties are required for each structured property or entry on different level regardless whether restoration is done in-flight or via caching, and restoring it in-flight can save the cost of caching and fits with the design of the reader.

---
In reply to: [233553959](https://github.com/OData/odata.net/pull/1106#discussion_r233553959) [](ancestors = 233553959,233299115,233190829)

Reply:
Again, I'm interested in whether we can avoid determining the set of nullable properties in the SelectedPropertiesNode each time, not evaluating what the set of actual null property values is for each resource.

---
In reply to: [233666008](https://github.com/OData/odata.net/pull/1106#discussion_r233666008) [](ancestors = 233666008,233553959,233299115,233190829)